### DISCUSSION TITLE
Реализовать более умный StreamedState.accept(), который акцептит только уникальные значения

### DISCUSSION THREAD

**Discussion by saturov** [surfstudio/SurfGear#179] [Ideas]
У нас из проекта в проект кочует extension, который фильтрует значения, переданные на accept в StreamedState и акцептит только те, которые отличаются от текущего. Это помогает избежать зацикливания процессов, которое может возникнуть при отсутствии такой проверки.

- Делать это поведением по-умолчанию не нужно, так как в некоторых случаях имеет смысл акцептить любые значения, даже идентичные.
- Также, предлагаю сделать это не в виде extension, а в виде дополнительной функции StreamedState (и всего семейства аналогичные классов).

```dart
extension AcceptUnique on StreamedState {
  void acceptUnique([dynamic data]) {
    if (value == data) {
      return;
    }
    accept(data);
  }
}
```

**Comment by r3tam** [comment]
А чем не устраивает distinct на стрим? RX же есть

**Reply by saturov** [reply]
Насколько я понимаю, в этом случае дублирующее значение физически всё равно будет в стриме, а это иногда не нужно.

**Reply by r3tam** [reply]
Это для синхронного получения данных? Но тогда какой смысл?

**Reply by r3tam** [reply]
просто кажется distinct вполне решает эту проблему

**Comment by pretorean** [comment]
а как быть если в StreamedState массив значений?

**Reply by saturov** [reply]
В этом случае точно такой же массив не будет заакцепчен. Ну а повторяемость элементов массива уже не на уровне работы со стримом проверять нужно, если ты об этом.

**Comment by HazarD2311** [comment]
Если будем это имплементить, то хотелось бы еще видеть, который все значения экшна будет унифицировать:
```
Action(distinctUnique: true)
```

**Reply by saturov** [reply]
Что это вообще значит, поясни?

**Comment by IlyaMax** [comment]
Может сделать как предлагает @HazarD2311 и для стейтов? Чтобы не было лишнего метода accept. По-моему не должно быть таких кейсов когда нужно одному state вызвать acceptUnique а потом accept.
